Make `#[used]` work when linking with `ld64`

madsmtm opened the pull request:
To make `#[used]` work in static libraries, we use the `symbols.o` trick introduced in https://github.com/rust-lang/rust/pull/95604.

However, the linker shipped with Xcode, ld64, works a bit differently from other linkers; in particular, [it completely ignores undefined symbols by themselves](https://github.com/apple-oss-distributions/ld64/blob/ld64-954.16/src/ld/parsers/macho_relocatable_file.cpp#L2455-L2468), and only consider them if they have relocations (something something atoms something fixups, I don't know the details).

So to make the `symbols.o` file work on ld64, we need to actually insert a relocation. That's kinda cumbersome to do though, since the relocation must be valid, and hence must point to a valid piece of machine code, and is hence very architecture-specific.

Fixes https://github.com/rust-lang/rust/issues/133491, see that for investigation.

---

Another option would be to pass `-u _foo` to the final linker invocation. This has the problem that `-u` causes the linker to not be able to dead-strip the symbol, which is undesirable. (If we did this, we would possibly also want to do it by putting the arguments in a file by itself, and passing that file via `@`, e.g. `@undefined_symbols.txt`, similar to https://github.com/rust-lang/rust/issues/52699, though that [is only supported since Xcode 12](https://developer.apple.com/documentation/xcode-release-notes/xcode-12-release-notes#Linking), and I'm not sure we wanna bump that).

Various other options that are probably all undesirable as they affect link time performance:
- Pass `-all_load` to the linker.
- Pass `-ObjC` to the linker (the Objective-C support in the linker has different code paths that load more of the binary), and instrument the binaries that contain `#[used]` symbols.
- Pass `-force_load` to libraries that contain `#[used]` symbols.

Failed attempt: Embed `-u _foo` in the object file with `LC_LINKER_OPTION`, akin to https://github.com/rust-lang/rust/issues/121293. Doesn't work, both because `ld64` doesn't read that from archive members unless it already has a reason to load the member (which is what this PR is trying to make it do), and because `ld64` only support the `-l`, `-needed-l`, `-framework` and `-needed_framework` flags in there.

---

TODO:
- [x] Support all Apple architectures.
- [x] Ensure that this works regardless of the actual type of the symbol.
- [x] Write up more docs.
- [x] Wire up a few proper tests.

@rustbot label O-apple

madsmtm committed:
Make #[used] work when linking with ld64

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12151162206/job/33885285956) [(plain)](https://github.com/rust-lang/rust/commit/c6251a8799f316ca09211f971c3305322a0e3c93/checks/33885285956/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.893 Building wheels for collected packages: reuse
#13 2.894   Building wheel for reuse (pyproject.toml): started
#13 3.142   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.143   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.144   Stored in directory: /tmp/pip-ephem-wheel-cache-_9minltx/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.146 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.554 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.554 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.103 Collecting virtualenv
#13 4.103 Collecting virtualenv
#13 4.146   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.254      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 40.5 MB/s eta 0:00:00
#13 4.294 Collecting distlib<1,>=0.3.7
#13 4.311   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.320      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 74.7 MB/s eta 0:00:00
#13 4.354 Collecting platformdirs<5,>=3.9.1
#13 4.361   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.402 Collecting filelock<4,>=3.12.2
#13 4.407   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.489 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.676 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      344000 kB
DirectMap2M:     8044544 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5703 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/run-make/include-all-symbols-linking/rmake.rs:10: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_codegen_ssa/src/back/link.rs:2115: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_codegen_ssa/src/back/link.rs:2117: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Wed Dec  4 01:32:08 UTC 2024
  network time: Wed, 04 Dec 2024 01:32:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

madsmtm commented on compiler/rustc_codegen_ssa/src/back/apple.rs:
```diff
@@ -26,6 +26,87 @@ pub(super) fn macho_platform(target: &Target) -> u32 {
+    symbol: object::write::SymbolId,
+    target: &Target,
+) -> object::write::Result<()> {
+    let (data, align, addend, r_type): (&[u8], _, _, _) = match &*target.arch {
+        "arm" => (&[0xff, 0xf7, 0xfe, 0xbf], 2, 0, object::macho::ARM_THUMB_RELOC_BR22),
+        "aarch64" => (&[0, 0, 0, 0x14], 4, 0, object::macho::ARM64_RELOC_BRANCH26),
+        "x86_64" => (
+            &[0x55, 0x48, 0x89, 0xe5, 0x5d, 0xe9, 0, 0, 0, 0],
+            16,
+            -4,
+            object::macho::X86_64_RELOC_BRANCH,
+        ),
+        "x86" => (&[0x55, 0x89, 0xe5, 0x5d, 0xe9, 0xf7, 0xff, 0xff, 0xff], 16, -4, 0),
+        arch => unimplemented!("unsupported Apple architecture {arch:?}"),
+    };
```
I'm quite unhappy with this, it _feels_ very brittle to emit a manual relocation like this, but I think it's the only way to do it (without pulling in a bunch of LLVM codegen internals).

I'm also unsure of the correctness, there might be more target options and/or flags that we should be checking here? Or maybe we should emit different kinds of relocations based on the kind of symbol?

thomcc replied:
Yeah... emitting manual relocations like this feels super fragile and hard to maintain. At the very least this would need comments and ideally references.

madsmtm replied:
Your suggestions here sounds correct to me, I've also later been thinking that an address offset instruction would be safer than a jump (since address offsets can be done safely on both function pointers and data pointers).

> You may want to add a local symbol

I'm not sure I understand what you mean by that? You want a local symbol that appears as the "entry" point for the section? Does that help the linker see that the code is unused?

madsmtm commented on compiler/rustc_codegen_ssa/src/back/apple.rs:
```diff
@@ -26,6 +26,87 @@ pub(super) fn macho_platform(target: &Target) -> u32 {
+        offset,
+        addend,
+        symbol,
+        flags: object::write::RelocationFlags::MachO { r_type, r_pcrel: true, r_length: 2 },
```
Note that we _could_ outsource part of this to `object` by using `object::write::RelocationFlags::Generic`, but we'd still have to write the machine code ourselves.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12553348093/job/35000563802) [(plain)](https://github.com/rust-lang/rust/commit/0653e4c2e94de45128018406009cab61a7da17c0/checks/35000563802/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.148 Building wheels for collected packages: reuse
#13 3.149   Building wheel for reuse (pyproject.toml): started
#13 3.411   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.412   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.412   Stored in directory: /tmp/pip-ephem-wheel-cache-9tahqqia/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.415 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.824 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.825 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.403 Collecting virtualenv
#13 4.403 Collecting virtualenv
#13 4.476   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.585      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 40.1 MB/s eta 0:00:00
#13 4.646 Collecting filelock<4,>=3.12.2
#13 4.661   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.699 Collecting platformdirs<5,>=3.9.1
#13 4.714   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.733 Collecting distlib<1,>=0.3.7
#13 4.755      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 82.4 MB/s eta 0:00:00
#13 4.755      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 82.4 MB/s eta 0:00:00
#13 4.837 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.025 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 5.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      116672 kB
DirectMap2M:     4077568 kB
DirectMap1G:    14680064 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5752 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_codegen_ssa/src/back/apple.rs:54: unexplained "```ignore" doctest; try one:
* make the test actually pass, by adding necessary imports and declarations, or
* make the test actually pass, by adding necessary imports and declarations, or
* use "```text", if the code is not Rust code, or
* use "```compile_fail,Ennnn", if the code is expected to fail at compile time, or
* use "```should_panic", if the code is expected to fail at run time, or
* use "```no_run", if the code should type-check but not necessary linkable/runnable, or
* explain it like "```ignore (cannot-test-this-because-xxxx)", if the annotation cannot be avoided.

removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Dec 31 00:08:40 UTC 2024
  network time: Tue, 31 Dec 2024 00:08:40 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

madsmtm commented:
@estebank I'm gonna mark this as ready for an initial review, to make sure that you agree that the approach here is the right one (vs. the alternatives I noted in the PR description / another alternative that I haven't considered). If so, then I'll complete the remaining TODO items.

estebank commented:
I'm sorry. I don't think I'm the best person to review this change.

r? compiler

jieyouxu commented:
I'm going to ask in T-compiler since this is likely to bounce.

DianQK commented:
I will try to review it tomorrow.
r? DianQK

DianQK commented:
IIUC, I can reproduce this with C:

```c
// foo.c
__attribute__((constructor))
void foo() {}

// main.c
int main() {
    return 0;
}
```

When I build the `foo.c` to the `libfoo.a`, `foo.o` that in `libfoo.a` never be loaded by any linker. For fix this, we added a `symbols.o` to help the linker load such object files.

I'm concerned about the current implementation. It seems we're building our logic around ld64's internal implementation details, and ld64 isn't truly open source. (I haven't looked into ld64's implementation details yet. I'm not sure if this approach is fragile or will be difficult to maintain in the future.)

IIUC, the key here is to **have the linker load object files as intended**, just like when we directly use `ld foo.o`. So why don't we just add these object files directly?

If we don't want to do this, how about adding the right value of `#[used]` to `symbols.o`? For `static PUSH: extern "C" fn() = push;`, we add "`push`" to `symbols.o`. That looks workable, as long as `push` always be a global symbol and both `PUSH` and `push` stay in the same object file.

DianQK commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2085,6 +2097,21 @@ fn add_linked_symbol_object(
+        // and hence this is all very architecture-specific.
+        //
+        // See the following for a few details on the design of ld64:
+        // https://github.com/apple-oss-distributions/ld64/blob/ld64-951.9/doc/design/linker.html
```
This link doesn't seem to mention the details of how ld64 resolves undefined symbols, especially regarding relocations. Should we add links to the relevant code locations?

madsmtm replied:
Yeah, it was the closest I could find, but actual links to the code would be a good idea. I'll try to find the relevant parts when I get the time.

Note though that I think the new linker, `-ld_prime`, [is not open source](https://forums.developer.apple.com/forums/thread/749558)? So this will only be a weak reference in any case.

madsmtm commented:
> [...]
> When I build the `foo.c` to the `libfoo.a`, `foo.o` that in `libfoo.a` never be loaded by any linker. For fix this, we added a `symbols.o` to help the linker load such object files.

That's correct.

> I'm concerned about the current implementation. It seems we're building our logic around ld64's internal implementation details, and ld64 isn't truly open source. (I haven't looked into ld64's implementation details yet. I'm not sure if this approach is fragile or will be difficult to maintain in the future.)

I agree that it is fragile, but I don't think it is as bad as it looks; consider that linkers are mostly backwards compatible, and that it works _now_ (whereas before it just didn't work at all, in no versions of ld64). I honestly don't think Apple is gonna change how this works, and if they do, we'll have the Xcode betas to fix it.

The only thing I can conceivably think of that might break is:
- If symbols were treated differently depending on their type (function vs. static).
- If the linker figures out that the data that the relocation refers to is never actually used.

I guess if we wanted to make it even more robust, we'd do something like:
```rust
// symbols.rs
extern crate crate1;
extern crate crate2;
extern crate crate3;

fn _used_symbols() {
    let _static1 = crate1::STATIC1;
    let _static2 = crate2::STATIC2;
    let _fn3 = crate3::fn3;
}

// rustc symbols.rs --emit=obj
```

That is, emit a label that refers to the block of code that touches all symbols, such that the linker cannot assume the section to be unused. The roughly equivalent could be done with `object-rs` like so:
```rust
file.add_symbol(write::Symbol {
    name: "_used_symbols".into(),
    value: 0,
    size: 0,
    kind: SymbolKind::Text,
    scope: SymbolScope::Dynamic,
    weak: false,
    section: write::SymbolSection::Section(section_id),
    flags: SymbolFlags::None,
});
```

Was that clear? I can try to go in more detail here if you want? Or try to conjure up some contrived assembly code, and consider how ld64 would have to link that now and in the future?

> IIUC, the key here is to **have the linker load object files as intended**, just like when we directly use `ld foo.o`. So why don't we just add these object files directly?

I can see two ways to do that:
- Avoid archives altogether, just pass each object file in every crate to the linker (or maybe combine the different object files generated by each codegen unit into a "`libfoo.o`").
- Link just the object files that use `#[used]`. Requires somehow extracting the relevant object files from the archive at link time, or maybe doing it while building the crate itself? E.g. create `libfoo.a` and `foo_used_symbols.o`.

The former is bad for link-time performance (the linker can skip a lot of work for archives that if can't for object files), and the latter is either also bad for perf, or makes the integration between `rustc` and Cargo even more complex than it already is.

> If we don't want to do this, how about adding the right value of `#[used]` to `symbols.o`? For `static PUSH: extern "C" fn() = push;`, we add "`push`" to `symbols.o`. That looks workable, as long as `push` always be a global symbol and both `PUSH` and `push` stay in the same object file.

Hmm, not sure I understand?

Note that we'd still want e.g. `#[used] #[link_section = "..."] static FOO: i32 = 2;` to be seen by the linker, and here there's no "inner" symbol to refer to?

bjorn3 commented:
> If the linker figures out that the data that the relocation refers to is never actually used.

The linker would have still pulled in the object file by that time. And `#[used]` doesn't actually have to prevent linker GC of the section. Only make sure that it isn't discarded before the linker has a chance of see the symbol and make it's own decision about whether it is considered to be a root for the linker GC. The unstable `#[used(linker)]` also forces the linker to keep the symbol, but that emits a separate flag on the symbol to prevent it from being deleted by the linker when the linker actually sees it.

dianqk reviewed (commented):
> Note that we'd still want e.g. `#[used] #[link_section = "..."] static FOO: i32 = 2;` to be seen by the linker, and here there's no "inner" symbol to refer to?

Ah, yeah! This also reminder me of something. Consider the following C code:

```c
__attribute__((constructor)) static void push() {}
```

or 

```c
static void push() {}

#if defined(__linux__)
__attribute__((section(".init_array")))
#elif defined(__APPLE__)
__attribute__((section("__DATA,__mod_init_func,mod_init_funcs")))
#endif
static void (*const PUSH)(void) = push;
```

**There are no global symbols here.**  I think **rustc should not modify local symbols into global symbols through `#[used]`** - in this case, it's the same situation as the C code above, where an object file has no global symbols.

IMO, loading object files through a "symbols.o" doesn't seem like a good idea. Looking at the longer term, I don't think we should continue to go this way.

Looking at this PR itself, I don't think we need to focus on ld64's internal implementation, since we're trying to simulate a standard object file. Based on my newly added comments, I think this approach is still quite complicated, and we need more workarounds to resolve it. What do you think?

madsmtm commented:
> There are no global symbols here. I think rustc should not modify local symbols into global symbols through #[used] - in this case, it's the same situation as the C code above, where an object file has no global symbols.

Hmm, I think I see what you mean, but is it really an issue? My understanding is vague here, but I think global vs. local symbols are linker-only concepts, are they not? Perhaps things differ when it comes to dynamic libraries, but I don't the symbol is marked as "exported" (?) , so the linker should not treat things any differently there either?

> IMO, loading object files through a "symbols.o" doesn't seem like a good idea. Looking at the longer term, I don't think we should continue to go this way.

I'm not sure I understand, how do you propose we solve this instead, then? In a way, `#[used]` _has_ to make the symbol global for it to be seen by the linker, right? That's what you _want_ when using `#[used]`.

bjorn3 commented:
(technically `#[used]` is equal to `#[used(compiler)]` only means the compiler will keep the function, not the linker will keep the function. the latter needs the unstable `#[used(linker)]`. That said, `#[used]` meaning `#[used(compiler)]` was a mistake IMO and we probably should make it `#[used(linker)]` in the future.)

DianQK commented:
> > There are no global symbols here. I think rustc should not modify local symbols into global symbols through #[used] - in this case, it's the same situation as the C code above, where an object file has no global symbols.
> 
> Hmm, I think I see what you mean, but is it really an issue? My understanding is vague here, but I think global vs. local symbols are linker-only concepts, are they not? Perhaps things differ when it comes to dynamic libraries, but I don't the symbol is marked as "exported" (?) , so the linker should not treat things any differently there either?
> > IMO, loading object files through a "symbols.o" doesn't seem like a good idea. Looking at the longer term, I don't think we should continue to go this way.
> 
> In a way, `#[used]` _has_ to make the symbol global for it to be seen by the linker, right? That's what you _want_ when using `#[used]`.

As @bjorn3 described, `#[used]` only indicates that the symbol should be kept, rather than making it as a global symbol. Let's look at an example where we don't need  `#[used]`. Note that both `FOO` and `foo` are local symbols here, and the object file cannot be loaded by an undefined symbol. The local and global symbols here describe the visibility of symbols within the object file.

```
$ cat main.rs
#![allow(dead_code)]

#[cfg_attr(target_os = "linux", link_section = ".init_array")]
#[cfg_attr(
    target_os = "macos",
    link_section = "__DATA,__mod_init_func,mod_init_funcs"
)]
static FOO: fn() = foo;
fn foo() {
    println!("foo");
}

fn main() {}

$ rustc main.rs && ./main
foo

$ nm main | grep -i foo
0000000100044010 s __ZN4main3FOO17h1b58eb28acf3c7f7E
0000000100000cf8 t __ZN4main3foo17h86e604b79f2f44edE
```

>  I'm not sure I understand, how do you propose we solve this instead, then?

My current thinking is to treat this PR as a working solution for now, while loading direct object file is a long-term improvement goal. Besides the test case, could you address my two concerns above?
1. Ensure that the instructions we added in `symbols.o` can be eliminated
2. Make these instructions potentially safer by using address offset instruction

bjorn3 commented:
> As @bjorn3 described, #[used] only indicates that the symbol should be kept, rather than making it as a global symbol. 

Not even kept by the linker. Just kept long enough for the linker to be able to see it at all. The linker is free to discard the symbol if it wants to. It just has to see it such that it can make the decision whether to discard it or not. This makes it pretty much useless for any purpose other than global constructors. Even the section merging for distributed slices is broken with lld because of this and requires `#[used(linker)]` instead AFAIK.

madsmtm commented:
> My current thinking is to treat this PR as a working solution for now, while loading direct object file is a long-term improvement goal.

Ah, cool! Thanks for the review and discussion so far, btw!

> Besides the test case, could you address my two concerns above?

Certainly (though it might take a little while for me to get back into it)!
@rustbot author

> This makes it pretty much useless for any purpose other than global constructors. 

I get your point that it is pretty much useless _on its own_, but you can very much make it useful by adding `no_dead_strip` (or similar for other platforms) to the linker section.

bjorn3 commented:
> but you can very much make it useful by adding no_dead_strip (or similar for other platforms) to the linker section.

Adding such a flag to the symbol is what the unstable `#[used(linker)]` does. Outside of Mach-O rustc doesn't expose any way to set it for the entire section. And ELF doesn't have a way to set it for the entire section as opposed to individual symbols afaik.

madsmtm commented:
I have now rewritten this to use a generic / absolute relocation in the `__data` section instead of an instruction relocation, which is much simpler and should be less error-prone. I have also added a proper test, and written some more documentation, so I believe this is:
@rustbot ready

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13380279118/job/37367516006) [(plain)](https://github.com/rust-lang/rust/commit/842e485ed263dcf6fb5adfef3e00c37853fcfcca/checks/37367516006/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.458 Building wheels for collected packages: reuse
#12 3.459   Building wheel for reuse (pyproject.toml): started
#12 3.698   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.699   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.699   Stored in directory: /tmp/pip-ephem-wheel-cache-qib60hxk/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.702 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 4.182 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 4.183 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.793 Collecting virtualenv
#12 4.793 Collecting virtualenv
#12 4.838   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.914      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 61.7 MB/s eta 0:00:00
#12 4.977 Collecting platformdirs<5,>=3.9.1
#12 4.981   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 5.004 Collecting distlib<1,>=0.3.7
#12 5.018      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 80.2 MB/s eta 0:00:00
#12 5.060 Collecting filelock<4,>=3.12.2
#12 5.065   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 5.065   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 5.180 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.419 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 5.5s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      135104 kB
DirectMap2M:     8253440 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.06s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5847 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_codegen_ssa/src/back/link.rs:2154: \
Use a single space after dots in comments.
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Mon Feb 17 23:51:58 UTC 2025
  network time: Mon, 17 Feb 2025 23:51:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:pushpin: Commit 5605039528ae034bb3dae8492b004fd580160c6a has been approved by `DianQK`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=DianQK 5605039528ae034bb3dae8492b004fd580160c6a -->
<!-- homu: {"type":"Approved","sha":"5605039528ae034bb3dae8492b004fd580160c6a","approver":"DianQK","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5605039528ae034bb3dae8492b004fd580160c6a with merge 3b73000dd498630c8289073ed9a069b5fc19f811...
<!-- homu: {"type":"BuildStarted","head_sha":"5605039528ae034bb3dae8492b004fd580160c6a","merge_sha":"3b73000dd498630c8289073ed9a069b5fc19f811"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13392799172/job/37411466399)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13392799172/job/37411466399","builder_name":"checks-actions"} -->

DianQK commented:
> > Note that we'd still want e.g. `#[used] #[link_section = "..."] static FOO: i32 = 2;` to be seen by the linker, and here there's no "inner" symbol to refer to?
> 
> Ah, yeah! This also reminder me of something. Consider the following C code:
> 
> ```c
> __attribute__((constructor)) static void push() {}
> ```
> 
> or
> 
> ```c
> static void push() {}
> 
> #if defined(__linux__)
> __attribute__((section(".init_array")))
> #elif defined(__APPLE__)
> __attribute__((section("__DATA,__mod_init_func,mod_init_funcs")))
> #endif
> static void (*const PUSH)(void) = push;
> ```
> 
> **There are no global symbols here.** I think **rustc should not modify local symbols into global symbols through `#[used]`** - in this case, it's the same situation as the C code above, where an object file has no global symbols.
> 
> IMO, loading object files through a "symbols.o" doesn't seem like a good idea. Looking at the longer term, I don't think we should continue to go this way.
> 
> Looking at this PR itself, I don't think we need to focus on ld64's internal implementation, since we're trying to simulate a standard object file. Based on my newly added comments, I think this approach is still quite complicated, and we need more workarounds to resolve it. What do you think?

I've put a pull request that describes my proposal. #137426

madsmtm commented:
It's unclear to me why the test failed, but I've skipped it on Windows for now, it's not the platform whose problems I'm trying to resolve here.

madsmtm commented:
The test run for `x86_64-msvc-1` completed successfully.

DianQK commented:
@bors r+ rollup=iffy

bors commented:
:hourglass: Testing commit b2024300840c3ef94a97273531bdb37d56d50fb5 with merge 2ad1dfb90b7929760ede23680cad3f99162e5b1c...
<!-- homu: {"type":"BuildStarted","head_sha":"b2024300840c3ef94a97273531bdb37d56d50fb5","merge_sha":"2ad1dfb90b7929760ede23680cad3f99162e5b1c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13506035278/job/37736126067)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13506035278/job/37736126067","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13506035278/job/37735728659) [(plain)](https://github.com/rust-lang-ci/rust/commit/2ad1dfb90b7929760ede23680cad3f99162e5b1c/checks/37735728659/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] unicode_ident test:false 0.245
error: could not compile `unicode-ident` (lib)

Caused by:
  process didn't exit successfully: `D:\a\rust\rust\build\bootstrap\debug\rustc D:\a\rust\rust\build\bootstrap\debug\rustc --crate-name unicode_ident --edition=2018 C:\Users\runneradmin\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\unicode-ident-1.0.16\src\lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debug-assertions=off --check-cfg cfg(docsrs,test) --check-cfg "cfg(feature, values())" -C metadata=a163a2dc94512ad5 -C extra-filename=-9e92e9f68ebc7032 --out-dir D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0-bootstrap-tools\release\deps -L dependency=D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0-bootstrap-tools\release\deps --cap-lints allow -Z binary-dep-depinfo` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)
[RUSTC-TIMING] byteorder test:false 0.247
[RUSTC-TIMING] once_cell test:false 0.310
[RUSTC-TIMING] windows_sys test:false 2.850
Build completed unsuccessfully in 0:01:25
```

</details>

madsmtm commented:
I suspect that's spurious, see https://github.com/rust-lang/rust/issues/133958. Could we try this again @DianQK?

DianQK commented:
@bors retry

bors commented:
:hourglass: Testing commit b2024300840c3ef94a97273531bdb37d56d50fb5 with merge c51b9b6d5234aa8e50c3b87784113a1af1af47cb...
<!-- homu: {"type":"BuildStarted","head_sha":"b2024300840c3ef94a97273531bdb37d56d50fb5","merge_sha":"c51b9b6d5234aa8e50c3b87784113a1af1af47cb"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13520677043/job/37792132034)
Approved by: DianQK
Pushing c51b9b6d5234aa8e50c3b87784113a1af1af47cb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"DianQK","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13520677043/job/37792132034"},"merge_sha":"c51b9b6d5234aa8e50c3b87784113a1af1af47cb"} -->

madsmtm commented:
Fairly sure the perf is noise, this affects macOS/Apple platforms, which benchmarks aren't run on (?). On other platforms, this should only amount to a single extra branch per `#[used]` symbol.

Kobzol commented:
The cycles/max RSS results are noise indeed, sorry for that (fix is in progress). Instruction counts look genuine, but they're green, so all good :laughing:
